common/sgx/eeid_verifier.c

```diff
@@ -0,0 +1,186 @@
+
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include <stdlib.h>
+#include <string.h>
+
+#ifdef OE_BUILD_ENCLAVE
+#include <openenclave/enclave.h>
+#else
+#include <openenclave/host.h>
+#endif
+
+#include <openenclave/attestation/sgx/eeid_verifier.h>
+#include <openenclave/attestation/sgx/verifier.h>
+#include <openenclave/bits/eeid.h>
+#include <openenclave/internal/raise.h>
+#include <openenclave/internal/sgx/plugin.h>
+#include <openenclave/internal/trace.h>
+
+static oe_result_t _eeid_verifier_on_register(
+    oe_attestation_role_t* context,
+    const void* config_data,
+    size_t config_data_size)
+{
+    OE_UNUSED(context);
+    OE_UNUSED(config_data);
+    OE_UNUSED(config_data_size);
+
+    oe_result_t result = OE_UNEXPECTED;
+
+    oe_verifier_t* sgx_verify = oe_sgx_plugin_verifier();
+    result = oe_register_verifier(sgx_verify, NULL, 0);
+    return result == OE_ALREADY_EXISTS ? OE_OK : result;
+}
+
+static oe_result_t _eeid_verifier_on_unregister(oe_attestation_role_t* context)
+{
+    OE_UNUSED(context);
+    oe_result_t result = OE_UNEXPECTED;
+
+    oe_verifier_t* sgx_verify = oe_sgx_plugin_verifier();
+    result = oe_unregister_verifier(sgx_verify);
+    if (result != OE_OK && result != OE_NOT_FOUND)
+        return result;
+
+    return OE_OK;
+}
+
+typedef struct _header
+{
+    uint32_t version;
+    oe_uuid_t format_id;
+    uint64_t data_size;
+    uint8_t data[];
+} header_t;
+
+static oe_result_t _eeid_verify_evidence(
+    oe_verifier_t* context,
+    const uint8_t* evidence_buffer,
+    size_t evidence_buffer_size,
+    const uint8_t* endorsements_buffer,
+    size_t endorsements_buffer_size,
+    const oe_policy_t* policies,
+    size_t policies_size,
+    oe_claim_t** claims,
+    size_t* claims_length)
+{
+    OE_UNUSED(context);
+    OE_UNUSED(policies);
+    OE_UNUSED(policies_size);
+
+    oe_result_t result = OE_UNEXPECTED;
+
+    if (!evidence_buffer || evidence_buffer_size == 0 || !endorsements_buffer ||
+        endorsements_buffer_size == 0)
+        return OE_INVALID_PARAMETER;
+
+    eeid_evidence_t* eeide = (eeid_evidence_t*)evidence_buffer;
+    oe_claim_t* sgx_claims = NULL;
+    size_t sgx_claims_size = 0;
+    uint8_t* sgx_evidence_buffer = eeide->data;
+    size_t sgx_evidence_buffer_size = eeide->eeid_sz;
```

Comment:
Line 83 also appears to assume that eeid_sz is already in the correct byte order, which looks suspiciously like a bug.   If the evidence buffer is transmitted across a network, how do you know the other end isn't a big-endian machine?   I would expect all multi-byte integer fields that can be passed across a network to be passed in network byte order, and have htonl/ntohl done on them to convert them to an integer.

Reply:
That's another excellent point! I didn't come across any byte-order related things so far; how is this done in other parts of the codebase?

Reply:
(Done for the EEID plugin, not for the rest of the infrastructure.)